### ISSUE TITLE
Required output files are not created for ragtag.py correct

### ISSUE THREAD

**Issue by code-acc** [malonge/RagTag#16]
Hi Michael,

I got "c_query_against_ref.paf" file as an output for the following command:
ragtag.py correct --debug -t 16 --aligner <PATH for minimap2> -u -o <out_dir> <HiCanu_assembly_dir> <reference_assembly_dir>
Can you please suggest what changes needs to be done in this command to get {query_prefix}.corrected.fasta as output?

Also, the output was not saved in the out_dir. Instead was saved in ragtag installation folder. Can you please suggest how to overcome this?

Next, I want to try the same command with raw read data for validation. I will add -R, --read-aligner and -T options to the above command and since my data is HiFi, I plan to provide the raw reads that came out of sequencer as it is without correction. Hope that's right. Or do you suggest different options for HiFi data?

Also, want to note that ref_asm is haploid and HiCanu_asm is diploid. Hope that won't cause a big difference in this analysis. I will redo it with HiCanu_asm as haploid once I have it.

Thank you,
Minal

**Comment by malonge** [comment]
Hi there,

I will answer your questions in order:

1. It looks like you are missing the PATH after `--aligner`. That option should be used to specify a path to either minimap2 or nucmer. Also, the reference and query assemblies (`<HiCanu_assembly_dir>` and  `<reference_assembly_dir>`) should be fasta files, not directories. Though I am a little confused. Are you getting an error? Can you share the log?

2. How did you install RagTag? To help with this, I will also need to see the log.

3. That is exactly right. Hifi reads can be used as error-corrected long reads.

4. RagTag does not do any false duplication removal, so I suggest you perform that step prior to RagTag, if it is necessary. RagTag will simply try to order and orient all input query sequences without any consideration for haplotypes.

Thanks

**Comment by code-acc** [comment]
Thanks for your quick response!

1. Sorry the command I wrote is misleading. I did provide minimap2 path and fasta files. Here is the complete command:
ragtag.py correct --debug -t 16 --aligner /data1/_software/minimap2/minimap2 -u -o /data1/CluHar/PacBio/Celtic_Sea_CCS_data/Assembly_comparisons/RagTag/CS4 /data1/CluHar/Genomes/Assembly_v2.0.2/Ch_v2.0.2.fasta /data1/CluHar/PacBio/Celtic_Sea_CCS_data/denovo/HiCanu_2.0/CS4/CS4.contigs.fasta

Please find the attached log file as well.

2. conda install -c bioconda ragtag. Maybe not necessary, but I should mention that - I had problems in installation because of python2 default usage. Department IT suggested deleting miniconda2 in order to install ragtag with miniconda3. Then I could use just ragtag.py and today suddenly ragtag.py gives error as "bash: ragtag.py: command not found...". Going to talk to IT tomorrow but it's just strange but let me know if you have any idea why that's so.

Get you on 3 and 4.

Thank you so much,
Minal

[c_query_against_ref.paf.log](https://github.com/malonge/RagTag/files/5048460/c_query_against_ref.paf.log)

**Comment by malonge** [comment]
Hi there, thanks for these details. I should have specified that the log I am looking for is what RagTag reports to stderr.

But the minimap2 log you provide shows that there was an error generating the alignments. If you can provide the ragtag log, we can figure out what went wrong.

Thanks

**Comment by malonge** [comment]
Also, you may consider using virtualenv for installation. Then it would be easy to install with `pip install RagTag`.

**Comment by malonge** [comment]
I would need to confirm this, but there may be a problem specifying a full path with `-o`. This is how I parse the path:

https://github.com/malonge/RagTag/blob/5df41f1388d0e2e14f4d023b65a67df8f6124c57/ragtag_correct.py#L394

The log will confirm

**Comment by code-acc** [comment]
Hi Michael,

Thanks for your reply! This was the only log file generated by ragtag. How do I generate ragtag log? Nothing was written on the screen when ragtag was finished and there was no error. Let me know if I should modify the command and run again to generate log file.
I am curious where do you spot an error in minimap2 log? There are bunch of warnings but don't see any error. 

Thanks for the installation answer. I will try this.

**Comment by malonge** [comment]
Hi there,

If you run the following command

```
ragtag.py correct --debug -t 16 --aligner /data1/_software/minimap2/minimap2 -u -o /data1/CluHar/PacBio/Celtic_Sea_CCS_data/Assembly_comparisons/RagTag/CS4 /data1/CluHar/Genomes/Assembly_v2.0.2/Ch_v2.0.2.fasta /data1/CluHar/PacBio/Celtic_Sea_CCS_data/denovo/HiCanu_2.0/CS4/CS4.contigs.fasta > ragtag.out 2> ragtag.err
```

then you can just share the `ragtag.err` file. That should show the exact command used to run minimap2. It should also show exactly where the output is being written to.

Thanks

**Comment by code-acc** [comment]
Thank you again! I will run this and send you the log file as soon as it is finished.

**Comment by code-acc** [comment]
HI Michael,

Please find the attached log file.

[ragtag_err.txt](https://github.com/malonge/RagTag/files/5062442/ragtag_err.txt)

**Comment by malonge** [comment]
Hi there,

So I think the problem might be that you are specifying a full path for the output directory with `-o`. Right now, ragtag expects just the output directory name and that directory will be made in the current working directory. That is why `-o` is a STR type and not a PATH type.

Perhaps you can verify this by rerunning your command put just with a non-path output directory name. Like this:

```
ragtag.py correct --debug -t 16 --aligner /data1/_software/minimap2/minimap2 -u -o CS4 /data1/CluHar/Genomes/Assembly_v2.0.2/Ch_v2.0.2.fasta /data1/CluHar/PacBio/Celtic_Sea_CCS_data/denovo/HiCanu_2.0/CS4/CS4.contigs.fasta
```

I think the user should be able to specify a PATH rather than a STR, so I will change this in the next patch regardless. But I also hope it fixes your particular issue so you can hopefully get some results in the meantime.

**Comment by code-acc** [comment]
Thank you so much! I have started the run now and will let you know as it gets done.

**Comment by code-acc** [comment]
Hi Michael,

I got the same output and error again. Attaching the log file here. Can you please see again?

Thanks,
Minal
[ragtag_err.txt](https://github.com/malonge/RagTag/files/5070438/ragtag_err.txt)

**Comment by malonge** [comment]
hmm

What is in `/home/mjamsandekar/bin/virtual_env/bin/CS4/c_query_against_ref.paf` ?

Is it a binary file? Does it look like a normal PAF file?

**Comment by malonge** [comment]
Also, when running ragtag, is your current working directory `/home/mjamsandekar/bin/virtual_env/bin/` ?

**Comment by code-acc** [comment]
Hi Michael,

So here is the screenshot for how paf file looks like. I am not sure if it's corrupt, expect chromosomes in the 6th column. I am running minimap2 -x asm5 command separately to see if I can replicate this paf file with minimap2 alone. And yes, the current working directory is /home/mjamsandekar/bin/virtual_env/bin/. Moving output from here to another folder.
![image](https://user-images.githubusercontent.com/57204282/90292039-bbea6880-de46-11ea-98ce-9b3238d3b9e7.png)

**Comment by malonge** [comment]
Yes that is definitely not right, though I can't think of what would cause that error. I'll comment more when you provide the output for your standalone minimap2 run.

**Comment by malonge** [comment]
The `-o` enhancement has been added in [v1.0.1](https://github.com/malonge/RagTag/releases/tag/v1.0.1). Now any path can be specified with `-o`. 

Any updates on the corrupted PAF file?

**Comment by code-acc** [comment]
Hi Michael,

I think the problem is related with how the current version (2.17-r974-dirty) of minimap2 is indexing the reference assembly. I used the same data on another server which has minimap2 v2.17-r963-dirty and it works fine - no warnings for indexing and mapping. So I installed ragtag on this server with the same command and it worked with correct output and finished in a few minutes versus another server was doing the whole minimap2 alignment with ragtag.py and was taking days! So yes, I think the problem is solved. I am so sorry that I raised an issue which is not a ragtag issue. Many thanks for helping me.

So glad to see this new version of ragtag!

Thank you,
Minal
